uitools/toolkitcpp/cpp/Esri/ArcGISRuntime/Toolkit/ArcGISAuthenticationController.cpp

```diff
@@ -0,0 +1,263 @@
+  });
+}
+
+/*!
+  \brief Destructor.
+ */
+ArcGISAuthenticationController::~ArcGISAuthenticationController() = default;
+
+ArcGISAuthenticationController* ArcGISAuthenticationController::create(QQmlEngine* qmlEngine, QJSEngine* /*jsEngine*/)
+{
+  static QPointer<ArcGISAuthenticationController> instance = new ArcGISAuthenticationController(qmlEngine);
+
+  if (!instance)
+    instance = new ArcGISAuthenticationController(qmlEngine);
+
+  return instance;
+}
+
+ArcGISAuthenticationController* ArcGISAuthenticationController::instance()
+{
+  return create(nullptr, nullptr);
+}
+
+void ArcGISAuthenticationController::handleArcGISAuthenticationChallenge(ArcGISAuthenticationChallenge* challenge)
+{
+  std::lock_guard<std::mutex> lock(m_mutex);
+  m_currentChallenge = challenge;
+  emit currentAuthenticatingHostChanged();
+
+  // first see if we can handle this with OAuth
+  const auto requestUrl = challenge->requestUrl();
+  for (OAuthUserConfiguration* userConfiguration : std::as_const(m_userConfigurations))
+  {
+    if (userConfiguration->canBeUsedForUrl(requestUrl))
+    {
+      OAuthUserCredential::createAsync(userConfiguration, this).then(this, [this](OAuthUserCredential* credential)
+      {
+        if (!m_currentChallenge)
+          return;
+
+        m_currentChallenge->continueWithCredential(credential);
+        m_currentChallenge->deleteLater();
+        m_currentChallenge = nullptr;
+      }).onFailed(this, [this](const ErrorException& e)
+      {
+        if (!m_currentChallenge)
+          return;
+
+        m_currentChallenge->continueAndFailWithError(e.error());
+        m_currentChallenge->deleteLater();
+        m_currentChallenge = nullptr;
+      });
+
+      return;
+    }
+  }
+
+  emit displayUsernamePasswordSignInView();
+}
+
+void ArcGISAuthenticationController::continueWithUsernamePassword(const QString& username, const QString& password)
+{
+  if (!m_currentChallenge)
+    return;
+
+  TokenCredential::createAsync(m_currentChallenge->requestUrl(),
+                               username,
+                               password,
+                               0).then(this, [this](TokenCredential* credential)
+  {
+    m_currentChallenge->continueWithCredential(credential);
+    m_currentChallenge->deleteLater();
+    m_currentChallenge = nullptr;
+  }).onFailed(this, [this](ErrorException e)
+  {
+    if ((++m_currentChallengeFailureCount) >= s_maxChallengeFailureCount)
+    {
+      m_currentChallengeFailureCount = 0;
+      m_currentChallenge->continueAndFailWithError(e.error());
+      m_currentChallenge->deleteLater();
+      m_currentChallenge = nullptr;
+      return;
+    }
+
+    emit currentChallengeFailureCountChanged();
+    emit displayUsernamePasswordSignInView();
+  });
+}
+
+void ArcGISAuthenticationController::respond(const QUrl& url)
+{
+  if (!m_currentOAuthUserLoginPrompt)
+    return;
+
+  m_currentOAuthUserLoginPrompt->respond(url);
+  m_currentOAuthUserLoginPrompt->deleteLater();
```

Comment:
Is it possible to use `unique_ptr` or it's required to use `deleteLater`? If no, is it documented?

Reply:
Either one works. Do you think std::unique_ptr would be preferable? We can remove any parent that's set on this and just use unique_ptr.